code/src/java/pcgen/util/fop/FopTask.java

```diff
@@ -0,0 +1,310 @@
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ * 
+ * Created on Jan 3, 2016, 9:14:07 PM
+ */
+package pcgen.util.fop;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Date;
+import javax.xml.transform.ErrorListener;
+import javax.xml.transform.SourceLocator;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.sax.SAXResult;
+import javax.xml.transform.stream.StreamSource;
+import org.apache.fop.apps.FOPException;
+import org.apache.fop.apps.FOUserAgent;
+import org.apache.fop.apps.Fop;
+import org.apache.fop.apps.FopFactory;
+import org.apache.fop.apps.MimeConstants;
+import org.apache.fop.render.Renderer;
+import pcgen.cdom.base.Constants;
+import pcgen.system.ConfigurationSettings;
+import pcgen.util.Logging;
+
+/**
+ * This class is used to generate pdf files from an xml source. There are two ways to define the
+ * source of the task: files or inputstreams. The output of this task can either be an OutputStream
+ * which you can point to a file, or a Renderer. The Renderer is used by print preview and for
+ * direct printing.
+ *
+ * @author Connor Petty <cpmeister@users.sourceforge.net>
+ */
+public class FopTask implements Runnable
+{
+
+	private static final TransformerFactory TRANS_FACTORY = TransformerFactory.newInstance();
+
+	private static FopFactory createFopFactory()
+	{
+		FopFactory fopFactory = FopFactory.newInstance();
+		fopFactory.setStrictValidation(false);
+
+		// Allow optional customization with configuration file
+		String configPath = ConfigurationSettings.getOutputSheetsDir() + File.separator + "fop.xconf";
+		Logging.log(Logging.INFO, "Checking for config file at " + configPath);
+		File userConfigFile = new File(configPath);
+		if (userConfigFile.exists())
+		{
+			Logging.log(Logging.INFO, "FOPHandler using config file "
+					+ configPath);
+			try
+			{
+				fopFactory.setUserConfig(userConfigFile);
+			}
+			catch (Exception e)
+			{
+				Logging.errorPrint("Problem with FOP configuration "
+						+ configPath + ": ", e);
+			}
+		}
+		return fopFactory;
+	}
+
+	private final StreamSource inputSource;
+	private final StreamSource xsltSource;
+	private final Renderer renderer;
+	private final OutputStream outputStream;
+
+	private StringBuilder errorBuilder = new StringBuilder(32);
+
+	private FopTask(StreamSource inputXml, StreamSource xsltSource, Renderer renderer, OutputStream outputStream)
+	{
+		this.inputSource = inputXml;
+		this.xsltSource = xsltSource;
+		this.renderer = renderer;
+		this.outputStream = outputStream;
+	}
+
+	public static FopTask newFopTask(InputStream inputXml, File xsltFile, OutputStream outputPdf)
+	{
+		return new FopTask(new StreamSource(inputXml), new StreamSource(xsltFile), null, outputPdf);

```

Comment:
xsltFile can be null since it would mean that the identity transformer should be used.